src/Generator/Generators/CSharp/CSharpMarshal.cs

```diff
@@ -124,6 +125,8 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
                     if (array.Type.IsPointerToPrimitiveType(PrimitiveType.Void))
                         supportBefore.WriteLineIndent("{0}[i] = new global::System.IntPtr({1}[i]);",
                             value, Context.ReturnVarName);
+                    else if (array.Type.Desugar().TryGetClass(out @class) && @class.IsRefType)
+                        supportBefore.WriteLineIndent("{0}[i] = {1}.__CreateInstance(*(({1}.Internal*)&({2}[i * sizeof({1}.Internal)])));", value, array.Type, Context.ReturnVarName);
```

Comment:
Actually, another small thing, use CreateInstanceIdentifier here instead of hardcoding the value.